contrib/win32/win32compat/fileio.c

```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
+	memcpy(buf + abs_chars, output, min(out_size, bufsiz - abs_chars));
+	ret = (ssize_t) min(out_size + abs_chars, bufsiz);
 
 cleanup:
 
+	if (reparse_buffer)
+		free(reparse_buffer);
+	if (handle != INVALID_HANDLE_VALUE)
+		CloseHandle(handle);
 	if (wpath)
 		free(wpath);
 	if (output)
 		free(output);
-	if (resolved_link)
-		free(resolved_link);
 
-	return (ssize_t) ret;
+	return (ssize_t)ret;
+}
+
+int
+fileio_symlink(const char *target, const char *linkpath)
+{
+	if (target == NULL || linkpath == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	DWORD ret = 0;
+	wchar_t *target_utf16 = utf8_to_utf16(resolved_path(target));
+	wchar_t *linkpath_utf16 = utf8_to_utf16(resolved_path(linkpath));
+	wchar_t *resolved_utf16 = _wcsdup(target_utf16);
+	if (target_utf16 == NULL || linkpath_utf16 == NULL || resolved_utf16 == NULL) {
+		errno = ENOMEM;
+		ret = -1;
+		goto cleanup;
+	}
+
+	/* Relative targets are relative to the link and not our current directory
+	 * so attempt to calculate a resolvable path by removing the link file name
+	 * leaving only the parent path and then append the relative link:
+	 * C:\Path\Link with Link->SubDir\Target to C:\Path\SubDir\Target
+	 */
+	if (!is_absolute_path(target)) {
+
+		/* allocate area to hold the total possible path */
+		free(resolved_utf16);
+		size_t resolved_len = (wcslen(target_utf16) + wcslen(linkpath_utf16) + 1);
+		resolved_utf16 = malloc(resolved_len * sizeof(wchar_t));
+		if (resolved_utf16 == NULL) {
+			errno = ENOMEM;
+			ret = -1;
+			goto cleanup;
+		}
+
+		/* copy the relative target to the end of the link's parent */
+		wcscpy_s(resolved_utf16, resolved_len, linkpath_utf16);
+		convertToBackslashW(resolved_utf16);
+		wchar_t * ptr = wcsrchr(resolved_utf16, L'\\');
+		if (ptr == NULL) wcscpy_s(resolved_utf16, resolved_len, target_utf16);
+		else wcscpy_s(ptr + 1, resolved_len - (ptr + 1 - resolved_utf16), target_utf16);
+	}
+
+	/* unlike other platforms, we need to know whether the symbolic link target is
+	 * a file or a directory.  the only way we can confidently do this is to
+	 * get the attributes of the target.  therefore, our symlink() has the
+	 * limitation of only creating symlink with valid targets
+	 */
+	WIN32_FILE_ATTRIBUTE_DATA attributes = { 0 };
+	if (GetFileAttributesExW(resolved_utf16, GetFileExInfoStandard, &attributes) == FALSE) {
+		errno = errno_from_Win32LastError();
+		ret = -1;
+		goto cleanup;
+	}
+
+	/* use the attribute of the file to determine the proper flag to send */
+	DWORD create_flags = (attributes.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) ?
+		SYMBOLIC_LINK_FLAG_DIRECTORY : 0;
+
+	/* symlink creation on earlier versions of windows were a privileged op
+ 	 * and then an option was added to create symlink using from an unprivileged
+ 	 * context so we try both operations, attempting privileged version first.
+	 * note: 0x2 = SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE
+	 */
+	if (CreateSymbolicLinkW(linkpath_utf16, target_utf16, create_flags) == 0) {
```

Comment:
If the linkpath already exists, then this call will fail and we leave don't override existing linkpath?

Reply:
Correct.  If the path already exists, it should fail.  This is consistent with the linux symlink() behavior.  I just tested to be sure.